## Stemming

#### Stemming is reduce words into base or root words . basically it remove or strip suffix and prefix from word to root word but most of times it not valid results. means its don't have meaning or its not valid which is not exist in dictionaries. It is faster but not valid.

In [1]:
import nltk

#### stemming is performed by algorithm its name is porter stemmer and but actually stemming is implemented by PorterStemmer class so we import one class from nltk.stem module called PorterStemmer.

#### porter stemmer(algorithm) defines rules for stemming and porterstemmer class actually implements the rules of stemming

In [8]:
from nltk.stem import PorterStemmer   # class

In [14]:
stemming_ = PorterStemmer()

#### We take some sample of data to perform stemming

In [15]:
words = ['studying', 'ninez', 'care','fairly', 'easily', 'run','running','runner','ran','runs','going','goes','gone','go','history','historical']

In [16]:
print(words)

['studying', 'ninez', 'care', 'fairly', 'easily', 'run', 'running', 'runner', 'ran', 'runs', 'going', 'goes', 'gone', 'go', 'history', 'historical']


#### This words are extending words(not root words) so we will perform stemming onit 

In [17]:
stemming_

<PorterStemmer>

#### stem is actually function which take only one attribute or word at single time as input so we will pass in loop and we give single word at single time on basis of for loop iteration

In [29]:
for word in words:
    print(word, '---->' , stemming_.stem(word ,to_lowercase = True))

studying ----> studi
ninez ----> ninez
care ----> care
fairly ----> fairli
easily ----> easili
run ----> run
running ----> run
runner ----> runner
ran ----> ran
runs ----> run
going ----> go
goes ----> goe
gone ----> gone
go ----> go
history ----> histori
historical ----> histor


#### After stemming , we get some results out of that some are correct by meaning and by spelling also like run , runner , ran ,gone  ,  go ,care but there are some words are not correct by spell and not correct by meaning also like studi , fairli , easili , goe , histori , histor

### Lemmatization

#### Lemmatization is reduce the words into meaningful base/root word on basis of its context.so we always prefer it because it gives wors into meaningful base / root word in its context which is valid or exist in dictionaries. It is slower but it gives more accurate

In [32]:
from nltk.stem import WordNetLemmatizer

#### WordNetLemmatizer is class

In [33]:
word_lemmatizer = WordNetLemmatizer()

#### so WordNtlemmatizer class contains function called lemmatize it take single word at a time so we pass this function in for loop and per iteration it take single word 

In [37]:
for word in words:
    print( word , '---->', word_lemmatizer.lemmatize(word))

studying ----> studying
ninez ----> ninez
care ----> care
fairly ----> fairly
easily ----> easily
run ----> run
running ----> running
runner ----> runner
ran ----> ran
runs ----> run
going ----> going
goes ----> go
gone ----> gone
go ----> go
history ----> history
historical ----> historical


##### We noticed all words are meaningful and In stemming goes is returned as 'goe' (which not having any meaning) and in lemmatization it returns as go.

#### Whereever changes are required they make changes it like goes ---> go as base word . wherever changes are not required they are not make any changes they taken as base words only

## Here We perform word tokenization , stemming and we want only those tokens which does not contains stopwords

#### We take text for processing

In [39]:
text = "Bhole Baba is absconding, after over a hundred people died in a stampede at his Hathras congregation. One response to this tragedy is to ask if this baba is ‘real’ or ‘fake’. But this is a false binary. All gurus or babas work on our emotions, not logic. Only, some abuse their power and collapse, some get institutionalised over generations as gods and prophets.And they do not flourish in India alone. In America, they are called the Charismatics who tell you how the Holy Spirit works miracles in human lives. Blind will see. Lame will walk. Poor will be prosperous. They are so powerful that they now fund politicians, who ensure their voice shapes legislation, and they receive tax benefits and protection from law. They have convinced Americans that Jesus wants them to carry guns, hate homosexuals, and deny women reproductive rights."

In [40]:
print(text)

Bhole Baba is absconding, after over a hundred people died in a stampede at his Hathras congregation. One response to this tragedy is to ask if this baba is ‘real’ or ‘fake’. But this is a false binary. All gurus or babas work on our emotions, not logic. Only, some abuse their power and collapse, some get institutionalised over generations as gods and prophets.And they do not flourish in India alone. In America, they are called the Charismatics who tell you how the Holy Spirit works miracles in human lives. Blind will see. Lame will walk. Poor will be prosperous. They are so powerful that they now fund politicians, who ensure their voice shapes legislation, and they receive tax benefits and protection from law. They have convinced Americans that Jesus wants them to carry guns, hate homosexuals, and deny women reproductive rights.


#### First we perform sentence tokenization  means text into individual sentences

In [96]:
from nltk.tokenize import sent_tokenize

In [121]:
sentences = sent_tokenize(text)

In [122]:
print(sentences)

['Bhole Baba is absconding, after over a hundred people died in a stampede at his Hathras congregation.', 'One response to this tragedy is to ask if this baba is ‘real’ or ‘fake’.', 'But this is a false binary.', 'All gurus or babas work on our emotions, not logic.', 'Only, some abuse their power and collapse, some get institutionalised over generations as gods and prophets.And they do not flourish in India alone.', 'In America, they are called the Charismatics who tell you how the Holy Spirit works miracles in human lives.', 'Blind will see.', 'Lame will walk.', 'Poor will be prosperous.', 'They are so powerful that they now fund politicians, who ensure their voice shapes legislation, and they receive tax benefits and protection from law.', 'They have convinced Americans that Jesus wants them to carry guns, hate homosexuals, and deny women reproductive rights.']


In [123]:
print(len(sentences))

11


In [124]:
from nltk.tokenize import word_tokenize  # word tokenization
from nltk.corpus import stopwords    # stopwords

#### We are ready with snetences

#### Now we perform word tokenization

In [127]:
for sentence in range(len(sentences)):
    
    word_tokens = word_tokenize(sentences[sentence])
    
    
    stem_sent = []
    
    for word in word_tokens:   # iteration on word tokenization 
        
        if word not in set(stopwords.words('english')): # set is for unique stop words
            # Here we want only those words from text data which are not stop words or not
            # present in stopwords
            
            stem_word = PorterStemmer().stem(word)   # single stem word
            #print(stem_word)
            
            stem_sent.append(stem_word)   # stem word append in list 
        #stem_sent is a list it contains tokens stem words and we are in for loop means it create
        #that much list equal to your length of sentence tokens (11) means it create 11 lists here
        
            
    #print(stem_sent)  
    sentences[sentence] = ' '.join(stem_sent)  # we join a list to a string and we are at 
    #out of inner for loop means it will join till outer for loop naturally  get exhaust means 
    #its len(11) i.e. 0 to 10. means it will add all lists to sentences variable and we track
    # it(index) on sentence (iterating variable).

In [128]:
print(sentences)

['bhole baba abscond , hundr peopl die stamp hathra congreg .', 'one respon tragedi ask baba ‘ real ’ ‘ fake ’ .', 'fal binari .', 'guru baba work emot , logic .', 'onli , abu power collap , get institutionali gener god prophets.and flourish india alon .', 'america , call charismat tell holi spirit work miracl human live .', 'blind see .', 'lame walk .', 'poor prosper .', 'power fund politician , ensur voic shape legisl , receiv tax benefit protect law .', 'convinc american jesu want carri gun , hate homosexu , deni women reproduct right .']


### Here We perform word tokenization , lammitization and we want only those tokens which does not contains stopwords

#### sentence tokenize over text data

In [136]:
sentences = sent_tokenize(text)

In [137]:
print(sentences)

['Bhole Baba is absconding, after over a hundred people died in a stampede at his Hathras congregation.', 'One response to this tragedy is to ask if this baba is ‘real’ or ‘fake’.', 'But this is a false binary.', 'All gurus or babas work on our emotions, not logic.', 'Only, some abuse their power and collapse, some get institutionalised over generations as gods and prophets.And they do not flourish in India alone.', 'In America, they are called the Charismatics who tell you how the Holy Spirit works miracles in human lives.', 'Blind will see.', 'Lame will walk.', 'Poor will be prosperous.', 'They are so powerful that they now fund politicians, who ensure their voice shapes legislation, and they receive tax benefits and protection from law.', 'They have convinced Americans that Jesus wants them to carry guns, hate homosexuals, and deny women reproductive rights.']


In [138]:
for sentence in range(len(sentences)):
    
    word_tokens = word_tokenize(sentences[sentence])  # word tokenize over sentence
    
    
    lemmatize_sent = []
    
    for word in word_tokens:   # iteration on word tokenization 
        
        if word not in set(stopwords.words('english')): # set is for unique stop words
            # Here we want only those words from text data which are not stop words or not
            # present in stopwords
            
            lemmatize_word = WordNetLemmatizer().lemmatize(word)   # single lemmatize word

            
            lemmatize_sent.append(lemmatize_word)   # lemmatize word append in list 
        #lemmatize_sent is a list it contains tokens lemmatize words and we are in for loop means 
        #it createthat much list equal to your length of sentence tokens (11) means it create 
        #11 lists here one list is made till your last word of sentences(sentence tokenize of 
        #per sentence) when your single sentence (by sentence_tokenizer) get naturally 
        #exhausted then it create for anotherlist till your all 11 sentences is not covered
        # in lemmatize sentence it append only those lemmatize word which are actually part of
        # that perticular lemmatize sentence eg : 1st lemmatize sentence contains only those 
        # words append which is part of 1st lammitizr sentence . in 2nd lemmatize sentence append
        # only those words whose actually part of 2nd lemmatize sentence. so like that it will
        # complete 11 lists or 11 lemmatize sentence.
        
    sentences[sentence] = ' '.join(lemmatize_sent)  # we join a list to a string and we are at 
    #out of inner for loop means it will join till outer for loop naturally  get exhaust means 
    #its len(11) i.e. 0 to 10. means it will add all lists to sentences variable and we track
    # it(index) on sentence (iterating variable).

In [139]:
print(sentences)

['Bhole Baba absconding , hundred people died stampede Hathras congregation .', 'One response tragedy ask baba ‘ real ’ ‘ fake ’ .', 'But false binary .', 'All guru baba work emotion , logic .', 'Only , abuse power collapse , get institutionalised generation god prophets.And flourish India alone .', 'In America , called Charismatics tell Holy Spirit work miracle human life .', 'Blind see .', 'Lame walk .', 'Poor prosperous .', 'They powerful fund politician , ensure voice shape legislation , receive tax benefit protection law .', 'They convinced Americans Jesus want carry gun , hate homosexual , deny woman reproductive right .']


In [140]:
print(len(sentences))

11


In [141]:
sentences

['Bhole Baba absconding , hundred people died stampede Hathras congregation .',
 'One response tragedy ask baba ‘ real ’ ‘ fake ’ .',
 'But false binary .',
 'All guru baba work emotion , logic .',
 'Only , abuse power collapse , get institutionalised generation god prophets.And flourish India alone .',
 'In America , called Charismatics tell Holy Spirit work miracle human life .',
 'Blind see .',
 'Lame walk .',
 'Poor prosperous .',
 'They powerful fund politician , ensure voice shape legislation , receive tax benefit protection law .',
 'They convinced Americans Jesus want carry gun , hate homosexual , deny woman reproductive right .']

### Regex

In [142]:
import re

In [143]:
input_str = "thhhhhe film Titanic wasssss released in 1988"

In [144]:
match_ = re.match(r'[A-Za-z]+' , input_str)
print(match_)

<re.Match object; span=(0, 7), match='thhhhhe'>


#### match : match is the finding or matching very first word with your pattern if your very first word matches with your pattern then it retrive the object means match object and index of match object

In [145]:
input_str= "1998 thhhhhe film Titanic wasssss released in"

In [146]:
match__ = re.match(r'[A-za-z]+' , input_str)
print(match__)

None


#### Becasue match object is not able to find the string in very first word

In [153]:
match___ = re.match(r'dog' , 'dog cat dog')

In [156]:
print(match___.group(0))

dog


In [158]:
print(match___.group())

dog


### Accessing corpus text - Mini Project

#### We perform a mini project on inbuilt data which is availabel in corpus library

In [160]:
import nltk
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

#### We import data from corpus library (it will show in pop up after nltk.downlaod()) so we will import movie_reviews ,product review_1 , product review_2 

In [205]:
from nltk.corpus import movie_reviews , product_reviews_1 , product_reviews_2

#### movie_reviews

In [206]:
print(len(movie_reviews.fileids()))

2000


#### product_reviews_1

In [207]:
print(len(product_reviews_1.fileids()))

6


#### product_reviews_2

In [208]:
print(len(product_reviews_2.fileids()))

10


#### That no . of text data into that reviews file

In [209]:
print(movie_reviews.fileids())

['neg/cv000_29416.txt', 'neg/cv001_19502.txt', 'neg/cv002_17424.txt', 'neg/cv003_12683.txt', 'neg/cv004_12641.txt', 'neg/cv005_29357.txt', 'neg/cv006_17022.txt', 'neg/cv007_4992.txt', 'neg/cv008_29326.txt', 'neg/cv009_29417.txt', 'neg/cv010_29063.txt', 'neg/cv011_13044.txt', 'neg/cv012_29411.txt', 'neg/cv013_10494.txt', 'neg/cv014_15600.txt', 'neg/cv015_29356.txt', 'neg/cv016_4348.txt', 'neg/cv017_23487.txt', 'neg/cv018_21672.txt', 'neg/cv019_16117.txt', 'neg/cv020_9234.txt', 'neg/cv021_17313.txt', 'neg/cv022_14227.txt', 'neg/cv023_13847.txt', 'neg/cv024_7033.txt', 'neg/cv025_29825.txt', 'neg/cv026_29229.txt', 'neg/cv027_26270.txt', 'neg/cv028_26964.txt', 'neg/cv029_19943.txt', 'neg/cv030_22893.txt', 'neg/cv031_19540.txt', 'neg/cv032_23718.txt', 'neg/cv033_25680.txt', 'neg/cv034_29446.txt', 'neg/cv035_3343.txt', 'neg/cv036_18385.txt', 'neg/cv037_19798.txt', 'neg/cv038_9781.txt', 'neg/cv039_5963.txt', 'neg/cv040_8829.txt', 'neg/cv041_22364.txt', 'neg/cv042_11927.txt', 'neg/cv043_16808.t

In [210]:
print(product_reviews_1.fileids())

['Apex_AD2600_Progressive_scan_DVD player.txt', 'Canon_G3.txt', 'Creative_Labs_Nomad_Jukebox_Zen_Xtra_40GB.txt', 'Nikon_coolpix_4300.txt', 'Nokia_6610.txt', 'README.txt']


In [211]:
print(product_reviews_2.fileids())

['Canon_PowerShot_SD500.txt', 'Canon_S100.txt', 'Diaper_Champ.txt', 'Hitachi_router.txt', 'Linksys_Router.txt', 'MicroMP3.txt', 'Nokia_6600.txt', 'README.txt', 'ipod.txt', 'norton.txt']


#### In movie reviews contains movie reviews data and other two contains product reviews

#### We take first product reviews data

#### We extract the data from norton.txt file

#### words is the function is reading predefined corpus text data

In [212]:
print(product_reviews_2.words('norton.txt'))

['I', 'have', 'read', 'the', 'installation', ...]


In [213]:
print(len(product_reviews_2.words('norton.txt')))

7259


##### We collect entire text data

In [214]:
entire_text = ' '.join(product_reviews_2.words('norton.txt'))

In [215]:
entire_text

'I have read the installation instructions for both NIS 2004 and NAV 2004 prior to installation , but still ended up with the same result ... junk software . Why is it that I can install any other type of software and it installs and works properly ? But if I installed either one of these Norton products , neither works after installation ? It can not be the computer or the owner , since I purchased McAfee Anti - Virus 8 and it installs and works fine with no problems . I have used Norton for the past 5 years and for the last 2 years , the software has gotten more and more disgraceful . I am glad that I do not work for Norton . I have used Norton products in the past and I am familiar with them . I bought NIS 2004 recently to try it out . After 4 attempted installs , I have finally given up . I am thankful that I have Roxio GOBACK , I had to use it 3 of the 4 times . During install , NIS would either freeze or not reboot properly when asked to do so . The MS scandisk would appear due t

#### We split this data into sentence tokens

In [216]:
from nltk.tokenize import sent_tokenize

In [217]:
norton_text_sent = sent_tokenize(entire_text)

In [218]:
print(len(norton_text_sent))

403


In [219]:
print(norton_text_sent)

['I have read the installation instructions for both NIS 2004 and NAV 2004 prior to installation , but still ended up with the same result ... junk software .', 'Why is it that I can install any other type of software and it installs and works properly ?', 'But if I installed either one of these Norton products , neither works after installation ?', 'It can not be the computer or the owner , since I purchased McAfee Anti - Virus 8 and it installs and works fine with no problems .', 'I have used Norton for the past 5 years and for the last 2 years , the software has gotten more and more disgraceful .', 'I am glad that I do not work for Norton .', 'I have used Norton products in the past and I am familiar with them .', 'I bought NIS 2004 recently to try it out .', 'After 4 attempted installs , I have finally given up .', 'I am thankful that I have Roxio GOBACK , I had to use it 3 of the 4 times .', 'During install , NIS would either freeze or not reboot properly when asked to do so .', '

### Here we perform lemmatization on this entire text data and before that we clean that data

In [220]:
print(len(norton_text_sent))

403


In [250]:
import time
start = time.time()
lemmatize_sent_ = []


for sentt in range(len(norton_text_sent)):   # norton_text_sent is sentence tokenize data

    
    mysent = re.sub(r'[^a-zA-Z]' , ' ' , norton_text_sent[sentt])
    #sub(substitute) is regex function searches the pattern with string if it matches then it
    #replace with replace value which we provide here we replace the non alphabates value by
    #space only. ^ is not and we are finding not alphabates over single letter
    # mysent contains only aplhabates data in sentence format
    mysent = mysent.lower()
    mysent = mysent.split() # split means we split the sentence on space means each word get
    # seperated in list
    # mysent is list now
    mysent = [WordNetLemmatizer().lemmatize(word) for word in mysent if  word not in set(stopwords.words('english'))]
    # now mysent is one list and it is coolection lemmatize words
    # Here we perform iteration on words because we already split mysent on space so then we don't
    #want stopwords so we are avoid it and we return lemmatize words in list only in mysent
    mysent = ' '.join(mysent)
    # mysent list we join in string and mysent is string now. means it is sentence in string format
    # that we want which is clean as well as lemmatize
    lemmatize_sent_.append(mysent)  # it append a single sentence into list
    
 
print(len(lemmatize_sent_))
print(time.time() - start)
print(lemmatize_sent_)
# It contains all 403 sentences because loop get iterate 403 times till naturally exhausted

403
2.642910957336426
['read installation instruction ni nav prior installation still ended result junk software', 'install type software installs work properly', 'installed either one norton product neither work installation', 'computer owner since purchased mcafee anti virus installs work fine problem', 'used norton past year last year software gotten disgraceful', 'glad work norton', 'used norton product past familiar', 'bought ni recently try', 'attempted installs finally given', 'thankful roxio goback use time', 'install ni would either freeze reboot properly asked', 'm scandisk would appear due manual shutdown would scan nd bar hr shut used goback', 'time install work error used unistall ni icon properly removed', 'never problem software computer', 'ni worked like charm last year', 'past norton customer consider slap face', 'buying software installed normally put position non working computer', 'norton downhill slid recently product one fallen cliff', 'loyal customer norton syman

### Here we perform stemmenting on this entire text data and before that we clean that data

In [235]:
print(len(norton_text_sent_))

403


In [246]:
import time
start = time.time()
stemm_sent_ = []


for sent in range(len(norton_text_sent)):   # norton_text_sent is sentence tokenize data

    
    mysent = re.sub(r'[^a-zA-Z]' , ' ' , norton_text_sent[sent])
   
    #sub(substitute) is regex function searches the pattern with string if it matches then it
    #replace with replace value which we provide here we replace the non alphabates value by
    #space only. ^ is not and we are finding not alphabates over single letter
    # mysent contains only aplhabates data in sentence format
    mysent = mysent.lower()
    mysent = mysent.split() # split means we split the sentence on space means each word get
    # seperated in list
    # mysent is list now
    mysent = [PorterStemmer().stem(word) for word in mysent if word not in set(stopwords.words('english'))]
    # now mysent is one list and it is coolection stemming words
    # Here we perform iteration on words because we already split mysent on space so then we don't
    #want stopwords so we are avoid it and we return stemmer words in list only in mysent
    mysent = ' '.join(mysent)
    # mysent list we join in string and mysent is string now. means it is sentence in string format
    # that we want which is clean as well as stemmed
    stemm_sent_.append(mysent)  # it append a single sentence into list
    
 
print(len(stemm_sent_))
print(time.time() - start)
print(stemm_sent_)
# It contains all 403 sentences because loop get iterate 403 times till naturally exhausted

403
2.581312656402588
['read instal instruct ni nav prior instal still end result junk softwar', 'instal type softwar instal work properli', 'instal either one norton product neither work instal', 'comput owner sinc purchas mcafe anti viru instal work fine problem', 'use norton past year last year softwar gotten disgrac', 'glad work norton', 'use norton product past familiar', 'bought ni recent tri', 'attempt instal final given', 'thank roxio goback use time', 'instal ni would either freez reboot properli ask', 'ms scandisk would appear due manual shutdown would scan nd bar hr shut use goback', 'time instal work error use unistal ni icon properli remov', 'never problem softwar comput', 'ni work like charm last year', 'past norton custom consid slap face', 'buy softwar instal normal put posit non work comput', 'norton downhil slid recent product one fallen cliff', 'loyal custom norton symantec product purchas internet secur softwar firewal could dissatisfi product henc one star rate', '

## Lemmatization is always gives best result.Lemmatization is slow but gives more accurate result and Stemming is fast but gives not valid result as compare to  Lemmatization .
## We noticed here Lemmatization take 2.64s and Stemming take 2.58 sec so lemmatization slower as compare to Stemming but more accurate than Stemming